modules/core/map/src/Element/FarmMap.php

```diff
@@ -50,6 +57,11 @@ public static function preRenderMap(array $element) {
     // Add the farm-map class.
     $element['#attributes']['class'][] = 'farm-map';
 
+    // By default, inform farm_map.js that it should instantiate the map.
+    if (empty($element['#attributes']['data-map-instantiator'])) {
```

Comment:
I think this could be moved to the `FarmMap::getInfo()` method - that is where "default" values can be defined, but can be overridden in a render array that uses `#type => 'farm_map'`

Reply:
More specifically, we could define `#attributes['data-map-instantiator'] = 'farm_map`` in the `getInfo()` method.

Reply:
Thanks for the review @paul121!

Am I right in thinking you're talking about something like the following?

```patch
diff --git a/modules/core/map/src/Element/FarmMap.php b/modules/core/map/src/Element/FarmMap.php
index 7f2b795f..8c1f7288 100644
--- a/modules/core/map/src/Element/FarmMap.php
+++ b/modules/core/map/src/Element/FarmMap.php
@@ -19,16 +19,19 @@ class FarmMap extends RenderElement {
   public function getInfo() {
     $class = get_class($this);
     return [
       '#pre_render' => [
         [$class, 'preRenderMap'],
       ],
       '#theme' => 'farm_map',
       '#map_type' => 'default',
+      '#attributes' => [
+        'data-map-instantiator' => 'farm_map',
+      ],
     ];
   }
 
   /**
    * Pre-render callback for the map render array.
    *
    * @param array $element
    *   A renderable array containing a #map_type property, which will be
@@ -52,21 +55,16 @@ class FarmMap extends RenderElement {
 
     // Get the map type.
     /** @var \Drupal\farm_map\Entity\MapTypeInterface $map */
     $map = \Drupal::entityTypeManager()->getStorage('map_type')->load($element['#map_type']);
 
     // Add the farm-map class.
     $element['#attributes']['class'][] = 'farm-map';
 
-    // By default, inform farm_map.js that it should instantiate the map.
-    if (empty($element['#attributes']['data-map-instantiator'])) {
-      $element['#attributes']['data-map-instantiator'] = 'farm_map';
-    }
-
     // Attach the farmOS-map and farm_map libraries.
     $element['#attached']['library'][] = 'farm_map/farmOS-map';
     $element['#attached']['library'][] = 'farm_map/farm_map';
 
     // Include map settings.
     $map_settings = !empty($element['#map_settings']) ? $element['#map_settings'] : [];
 
     // Include the map options.
```

I took a glance over some of the [core `ElementInterface::getInfo` implementations](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Render%21Element%21ElementInterface.php/function/overrides/ElementInterface%3A%3AgetInfo/9.0.x) and didn't find any which directly define default attributes as such.

I believe that is because the fields from `ElementInterface::getInfo` [are not deeply merged](https://www.drupal.org/project/drupal/issues/2070131) with values provided in the render array.

Instead, it appears that the convention is something like this;

```patch
diff --git a/modules/core/map/src/Element/FarmMap.php b/modules/core/map/src/Element/FarmMap.php
index 7f2b795f..3029129d 100644
--- a/modules/core/map/src/Element/FarmMap.php
+++ b/modules/core/map/src/Element/FarmMap.php
@@ -19,16 +19,17 @@ class FarmMap extends RenderElement {
   public function getInfo() {
     $class = get_class($this);
     return [
       '#pre_render' => [
         [$class, 'preRenderMap'],
       ],
       '#theme' => 'farm_map',
       '#map_type' => 'default',
+      '#data-map-instantiator' => 'farm_map',
     ];
   }
 
   /**
    * Pre-render callback for the map render array.
    *
    * @param array $element
    *   A renderable array containing a #map_type property, which will be
@@ -52,20 +53,19 @@ class FarmMap extends RenderElement {
 
     // Get the map type.
     /** @var \Drupal\farm_map\Entity\MapTypeInterface $map */
     $map = \Drupal::entityTypeManager()->getStorage('map_type')->load($element['#map_type']);
 
     // Add the farm-map class.
     $element['#attributes']['class'][] = 'farm-map';
 
-    // By default, inform farm_map.js that it should instantiate the map.
-    if (empty($element['#attributes']['data-map-instantiator'])) {
-      $element['#attributes']['data-map-instantiator'] = 'farm_map';
-    }
+    Element::setAttributes($element, [
+      'data-map-instantiator',
+    ]);
 
     // Attach the farmOS-map and farm_map libraries.
     $element['#attached']['library'][] = 'farm_map/farmOS-map';
     $element['#attached']['library'][] = 'farm_map/farm_map';
 
     // Include map settings.
     $map_settings = !empty($element['#map_settings']) ? $element['#map_settings'] : [];
 
```

Naively, that convention doesn't seem like it provides much benefit. It means that each "configurable attribute" now requires two methods to be changed instead of one. Is there some other functional benefit from having properties specified in the `ElementInterface::getInfo` method? The docs seem pretty sparse about the actual purpose/intent of most of this stuff...

Reply:
> Am I right in thinking you're talking about something like the following?

Yes, exactly what I was thinking. But shoot, good find! Looks like that won't work because of the deep merge issue.

> Naively, that convention doesn't seem like it provides much benefit.

I agree. It would be confusing to have both the top level `'#data-map-instantiator'` and  `#attributes['#data-map-instantiator']`. Keeping as is with just the latter make sense!